docs/rfc/active/data-permissions/README.md

```diff
@@ -0,0 +1,138 @@
+- Start Date: 2022-04-19
+- RFC PR: #4694
+- Discussion Issue: 
+- Implementation PR(s): 
+
+# Modeling Data Permissions
+
+## Summary
+
+Many datahub sources provide access control mechanisms to manage user access to assets.
+Datahub should provide functionality for viewing these permissions.
+
+## Motivation
+
+Within many sql systems, the data control language provides commands like GRANT, DENY and REVOKE to manage control to datasets.
+
+Being able to view grants is the first step toward supporting larger integrations with dataset permissions. [See Here](https://feature-requests.datahubproject.io/b/feedback/p/access-control-on-datasets-bi-assets)
+
+Of the datahub integration sources listed, the following have some form of data access control management:
+
+- Athena (via IAM)
+- BigQuery
+- Clickhouse
+- Druid
+- Elastic Search
+- Glue
+- Hive (via Apache Ranger)
+- Kafka
+- Looker
+- MariaDB
+- Metabase
+- MongoDB
+- MicrosoftSQLServer
+- MySQL
+- Oracle
+- PostgreSQL
+- Redash
+- Redshift
+- S3 (via IAM)
+- Sagemaker
+- Snowflake
+- Superset
+- Tableau
+- Trino
+
+## Requirements
+
+The permissions aspect must be:
+
+- Appliable to datasets, containers, dashboards
+- Support ingesting and viewing structured access control (GRANT {permission} ON {object} TO {user or role}
+- Support ingesting and viewing unstructured access control (IAM Policy Documents, Custom Policies from other datasets)
+-
+
+### Extensibility
+
+> Please also call out extensibility requirements. Is this proposal meant to be extended in the future? Are you adding
+> a new API or set of models that others can build on in later? Please list these concerns here as well.
+
+## Non-Requirements
+
+This RFC specifically calls for functionality to view the permissions. Actions taken on the permissions may be great areas for future enhancement but is not a focus here.
+
+## Detailed design
+
+There are no new Entities associated with this design.
+
+There is one aspect proposed:
+
+```
+record DataPermission {
+  structuredPermissions: array[StructuredDataPermission]
+  unstructuredPermissions: array[UnstructuredDataPermission]
+  relatedPermissions: array[Urn]
+}
+
+record StructuredDataPermission {
+   permission: string
+   object: Urn // if not specified, is implicitly the urn of the containing Entity
+   subject: union[corpUser, corpGroup, string] // grant to a user, group, or non-standard entity (e.g. "*", "all", "NONE")
+}
+
+record UnstructuredDataPermission {
+  language: string // maybe ENUM? (e.g. "IAM")
+  value: string
+}
+```
```

Comment:
Below is an outline that might allow us to start with the functionality and concepts scoped in this RFC/model, and then further evolve in multiple directions in future.

We could create a new Entity called _DataPolicy_ ; it has one primary aspect called _DataPolicyInfo_, something along the lines of:

    record DataPolicyInfo {
        type: DataPolicyType //"ResourcePrincipalPolicy", "MaskingPolicy" etc.

        resourcePrincipalPolicy: ResourcePrincipalPolicy

        // can evolve to add other types of policies here
    }

Note that each “policy” here is modeled as the most granular clause that can be specified. (Some third-party systems may allow specifying multiple clauses in a single “statement”.)

Each type of policy would contain reference to principal (user/group/role) and resource.

For example, _ResourcePrincipalPolicy_ would be something along the lines of:

    record ResourcePrincipalPolicy {
        principal: Urn // of CorpUser, CorpGroup or CorpRole

        resourceRef: ResourceReference // resource(s) on which policy is specified, see below

        permission: string  // native permission string e.g. "select" etc.; standard values could be identified for datahub
        isAllow: boolean    // is access allowed vs denied
    }

_ResourceReference_ is intended to capture various mechanisms in which (a) resource(s) can be referenced, something along the lines of:

    record ResourceReference {
        type: ResourceIdentifierType // "RESOURCE_URN", "SPECIAL" etc.
        resourceUrn: optional Urn    // dataset, container etc. as appropriate
        special: optional SpecialResource  // enum with "ALL" etc.

        // can evolve to add other ways of referencing resources e.g. with policy tags (special glossary terms)
    }

With a structure like the above, we would be able to capture ACLs as originally intended in this RFC. 
Additionally, we would also capture RBAC, and can evolve to add tag-based policies etc.

Also, all polices involving a resource or user would be query-able; one can envision a UI where all data-policies pertaining to users/resources are shown on their respective pages.

Notes:
- Relationship annotations on the fields like _principal_, _resourceUrn_ etc. would allow navigation across the entities.

- For RBAC and tag-based policies to be captured realistically, CorpRole would also need to be modeled; the typical user - group - role triad would be modeled in CorpUser - CorpGroup - CorpRole respectively.

- Purely native policies, i.e. "UnstructuredDataPermission" in original RFC, are not captured in this outline; they can be modeled specifically within certain policyTypes as required.